scene/gui/color_mode.h

```diff
@@ -0,0 +1,149 @@
+/* Permission is hereby granted, free of charge, to any person obtaining */
+/* a copy of this software and associated documentation files (the       */
+/* "Software"), to deal in the Software without restriction, including   */
+/* without limitation the rights to use, copy, modify, merge, publish,   */
+/* distribute, sublicense, and/or sell copies of the Software, and to    */
+/* permit persons to whom the Software is furnished to do so, subject to */
+/* the following conditions:                                             */
+/*                                                                       */
+/* The above copyright notice and this permission notice shall be        */
+/* included in all copies or substantial portions of the Software.       */
+/*                                                                       */
+/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,       */
+/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF    */
+/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.*/
+/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY  */
+/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,  */
+/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE     */
+/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                */
+/*************************************************************************/
+
+#ifndef COLOR_MODE_H
+#define COLOR_MODE_H
+
+#include "core/math/color.h"
+#include "core/templates/vector.h"
+#include "scene/gui/color_picker.h"
+
+class ColorMode {
+public:
+	ColorPicker *color_picker;
+
+	virtual String get_name() const = 0;
+
+	virtual int get_slider_count() const { return 3; };
+	virtual float get_slider_step() const = 0;
+	virtual String get_slider_label(int idx) const = 0;
+	virtual float get_slider_max(int idx) const = 0;
+	virtual double get_slider_value(int idx) const = 0;
+
+	virtual Color get_color() const = 0;
+
+	virtual void slider_draw(int p_which) = 0;
+	virtual bool apply_theme() const { return false; }
+	virtual ColorPicker::PickerShapeType get_shape_override() const { return ColorPicker::PickerShapeType::SHAPE_MAX; }
+
+	ColorMode(ColorPicker *p_color_picker);
+	virtual ~ColorMode(){};
+};
+
+class ColorModeHSV : public ColorMode {
+public:
+	String labels[3] = { "H", "S", "V" };
+	float slider_max[4] = { 359, 100, 100, 255 };
+
+	virtual String get_name() const override { return "HSV"; }
+
+	virtual float get_slider_step() const override { return 1.0; }
+	virtual String get_slider_label(int idx) const override;
+	virtual float get_slider_max(int idx) const override;
+	virtual double get_slider_value(int idx) const override;
+
+	virtual Color get_color() const override;
+
+	virtual void slider_draw(int p_which) override;
+
+	ColorModeHSV(ColorPicker *p_color_picker) :
+			ColorMode(p_color_picker){};
+
+	~ColorModeHSV(){};
```

Comment:
Empty constructors and destructors can be removed.

Reply:
Constructors can't be removed.

Reply:
Ah ok then.